test/e2e/autoscaler.go

```diff
@@ -226,6 +245,46 @@ func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput)
 			WaitForMachineDeployment: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
 		})
 
+		if input.ScaleToAndFromZero {
+			By("Enabling autoscaler for the MachineDeployment to zero")
+			// Enable autoscaler on the MachineDeployment.
+			framework.EnableAutoscalerForMachineDeploymentTopologyAndWait(ctx, framework.EnableAutoscalerForMachineDeploymentTopologyAndWaitInput{
+				ClusterProxy:                input.BootstrapClusterProxy,
+				Cluster:                     clusterResources.Cluster,
+				NodeGroupMinSize:            "0",
+				NodeGroupMaxSize:            mdNodeGroupMaxSize,
+				WaitForAnnotationsToBeAdded: input.E2EConfig.GetIntervals(specName, "wait-autoscaler"),
+			})
+
+			By("Scaling the MachineDeployment scale up deployment to zero")
+			framework.ScaleScaleUpDeploymentAndWait(ctx, framework.ScaleScaleUpDeploymentAndWaitInput{
+				ClusterProxy: workloadClusterProxy,
+				Replicas:     mpOriginalReplicas + 0,
+			}, input.E2EConfig.GetIntervals(specName, "wait-autoscaler")...)
+
+			By("Checking the MachineDeployment finished scaling down to zero")
+			framework.AssertMachineDeploymentReplicas(ctx, framework.AssertMachineDeploymentReplicasInput{
+				Getter:                   input.BootstrapClusterProxy.GetClient(),
+				MachineDeployment:        clusterResources.MachineDeployments[0],
+				Replicas:                 0,
+				WaitForMachineDeployment: input.E2EConfig.GetIntervals(specName, "wait-controllers"),
+			})
+
+			By("Scaling the MachineDeployment scale up deployment to 1")
+			framework.ScaleScaleUpDeploymentAndWait(ctx, framework.ScaleScaleUpDeploymentAndWaitInput{
+				ClusterProxy: workloadClusterProxy,
+				Replicas:     mpOriginalReplicas + 1,
```

Comment:
Same here, is this always 1, why MP?

Reply:
For CAPV this will always be 1.

If a provider supports both MP and MD, we need to add the number of existing machine pool nodes because the deployment can be scheduled otherwise and won't trigger a scale-up.

Reply:
Thx! Let's add a godoc comment

Reply:
From the KEP for scale from zero:

> Support for MachinePools, either with the cluster autoscaler or using infrastructure provider native implementations (eg AWS AutoScalingGroups).

[[0]](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md#machineset-and-machinedeployment-annotations)

Reply:
Removed it

Reply:
Yeah wanted to paste it at the other